### PR TITLE
Feature/field data map

### PR DESCRIPTION
**Pull request by maxgrossman**
in reference to #71 

this pr:

1. adds new route to visualize (map) road geometry by source (RouteShoot and RoadLabPro)
  - this includes new, related actions/reducers/store to fetch and add to state road geometries
  - new components (field-data-map.js / field-data-map-legend.js) to visualize geometries
  - related additions to sass for stying
  - a new util called field-data.js that generates the mapboxgl source & layers to show both sources
  - inclusion of util (currently in another pr) made for search functionality to generate a centerpoint and zoom to initialize the map with when given a bbox

2. adds new columns to aa-table-vpromms.js component that allows for: 
  - field data geojsons
  - link to field data analytics pages (currenlty only showing a map of the field data)

Achieving 2. required:
  - including new actions/reducers (fetchVProMMsIdsSources/VProMMsIdsSources) to fetch data from the api that tells whether or not field data exists in the database.

### CODE REVIEW

File: app/assets/scripts/actions/action-creators.js [hunk 1]
```diff
@@ -395,6 +395,83 @@ export function fetchVProMMSids () {
...
+export function fetchVProMMsidSourceGeoJSON (vprommId, provinceName) {
+  return function (dispatch) {
+    dispatch(requestVProMMSidSourceGeoJSON());
+    // hit the grouped field geometries endpoint. do not download it.
+    let url = `${config.api}/field/${vprommId}/geometries?grouped=true`;
+    return fetch(url)
+    .then(response => response.json())
+    .then(json => {
+      if (json.statusCode === 400) {
+        return { dataAvailable: false };
```
Comment (danielfdsilva) [inline app/assets/scripts/actions/action-creators.js]: This is not returned to anything

File: app/assets/scripts/components/aa-field-map.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    return (
+      <div>
+        <div id='aa-map' className='aa-map'></div>
+        <AAFieldMapLegend layers={layers} />
+      </div>
+    );
+  },
+
+  renderHeader: function (provinceName) {
+    return (<h1>{`${provinceName} Province - Road # ${this.props.roadId}`}</h1>);
```
Comment (danielfdsilva) [inline app/assets/scripts/components/aa-field-map.js]: Any specific reason why the content is a string?

File: app/assets/scripts/components/aa-table-vpromms.js [hunk 3]
```diff
@@ -75,17 +82,32 @@ const AATable = React.createClass({
...
+
+  makeFieldMapLink: function (vprommExists, roadId) {
+    const provinceId = this.props.province;
+    return vprommExists ? (
+      <Link to={`${getLanguage()}/analytics/${provinceId}/${roadId}/`}>{roadId}</Link>
+    ) : roadId;
+  },
+
+  // in renderTableBody, vrpommExists represents if a given vpromms id has a field data source attached to it.
+  // it used to decide whether to make a link to the AAFieldMap component in the first column and to a road data dump in the last column
```
Comment (danielfdsilva) [inline app/assets/scripts/components/aa-table-vpromms.js]: It's a good practice to keep comments at 80 characters per line.

File: app/assets/scripts/views/analytics-admin-area.js [hunk 4]
```diff
@@ -14,42 +15,85 @@ var AnalyticsAA = React.createClass({
...
+  },
+
+  // the component can either render the admin area's table, or it renders a vpromms road specific analytics page.
+  // if the current params does not include vpromms, then the table should be rendered.
+  // Or, if the vpromms exists, the vpromms analytics page needs to be rendered
+  render: function () {
+    const vprommsParam = this.props.params.vpromm;
+    if (!vprommsParam) {
+      return this.renderTable();
+    } else if (vprommsParam) {
```
Comment (danielfdsilva) [inline app/assets/scripts/views/analytics-admin-area.js]: This is a normal `else`no?

File: app/assets/scripts/views/editor.js [hunk 5]
```diff
@@ -76,6 +77,7 @@ var Editor = React.createClass({
   },
 
   render: function () {
+    console.log(this.props);
```
Comment (danielfdsilva) [inline app/assets/scripts/views/editor.js]: stray console.log

File: app/assets/scripts/views/editor.js [hunk 6]
```diff
@@ -46,8 +47,8 @@ var Editor = React.createClass({
       switch (e.data.id) {
         case 'or-editor':
           var hash = this.cleanUrl(e.data.url, config.editorUrl);
-          this.props.dispatch(replace(`/editor/${hash}`));
           this.props._setGlobalZoom(hash);
+          this.props.dispatch(replace(`/${getLanguage()}/editor/${hash}`));
```
Comment (danielfdsilva) [inline app/assets/scripts/views/editor.js]: This would be better if passed as a prop. No strong feelings though

**Comment by maxgrossman** [comment]
@olafveerman - can you also give this a look through since you are familiar with the api routes power this new feature?

also @ricardomestre - just flagging this for you so you are aware that new styling will be needed.

**Comment by olafveerman** [comment]
@maxgrossman Can you address the merge conflicts and check for stray `console.log`'s?
Also, since the new feature feels very much a part of the dashboards, can we have a route that includes `/analytics`? eg. `/en/analytics/road/214YD00061`?

@danielfdsilva will do a review after the conflicts are solved.

File: app/assets/scripts/actions/action-creators.js [hunk 7]
```diff
@@ -401,6 +400,81 @@ export function fetchVProMMsids () {
...
+export function fetchVProMMSidsSources (ids) {
+  return function (dispatch) {
+    dispatch(requestVProMMSidsSources());
+    let idsQuery = ids.join(',');
+    let url = `${config.api}/field/${idsQuery}/exists`;
+    return fetch(url)
+    .then(response => response.json())
+    .then(json => {
+      if (json.statusCode === 400) {
+        return {dataAvailable: false};
```
Comment (danielfdsilva) [inline app/assets/scripts/actions/action-creators.js]: This is not returned to anything

File: app/assets/scripts/components/aa-field-map-legend.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    sources: React.PropTypes.array
+  },
+
+  renderLegendElements: function () {
+    // for each source as the source name
+    // as well as a small svg path with the color
+    // matching source in the lineColors config
+    return this.props.sources.map((source, i) => {
+      return (
+        <li key={i}>
```
Comment (danielfdsilva) [inline app/assets/scripts/components/aa-field-map-legend.js]: It's not a good idea to use the `index` as a component key. Is there anything else that could be used? `source` perhaps?

File: app/assets/scripts/components/aa-field-map.js [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    return (
+      <div>
+        <div id='aa-map' className='aa-map'></div>
+        <AAFieldMapLegend sources={uniq(sources)} />
+      </div>
+    );
+  },
+
+  renderHeader: function (adminName) {
+    return (<h1>{`${adminName} Province - Road # ${this.props.params.vpromm}`}</h1>);
```
Comment (danielfdsilva) [inline app/assets/scripts/components/aa-field-map.js]: No need to use template literals

File: app/assets/scripts/reducers/reducer.js [hunk 10]
```diff
@@ -364,6 +364,63 @@ const VProMMsWayBboxDefaultState = {
...
+      state.geoJSON = action.json;
+      state.vprommId = action.vprommId;
+      state.provinceName = action.provinceName;
+      state.fetching = false;
+      state.fetched = true;
+      break;
+    case actions.REMOVE_VPROMMS_SOURCE_GEOJSON:
+      console.log('REMOVE_SOURCE_GEOJSON');
+      state = _.cloneDeep(state);
+      state = defaultVProMMSidSourceGeoJSON;
```
Comment (danielfdsilva) [inline app/assets/scripts/reducers/reducer.js]: No need to clone state, you can just return the default one

File: app/assets/scripts/main.js [hunk 11]
```diff
@@ -65,7 +66,10 @@ render((
         <Route path='editor/*' component={Editor} pageClass='editor' />
         <Route path='explore' component={Explore} pageClass='explore' />
         <Route path='analytics' component={Analytics}>
-          <IndexRoute component={AnalyticsIndex} pageClass='analytics' />
+          <Route path='road' component={Analytics} >
+            <Route path=':vpromm' component={AAFieldMap} pageClass='analytics-aa'/>
+          </Route>
+          <Route path='main' component={AnalyticsIndex} pageClass='analytics' />
```
Comment (danielfdsilva) [inline app/assets/scripts/main.js]: You should be using a IndexRoute here. Looks better than /main.
Any reason you went with this approach?

File: app/assets/scripts/reducers/reducer.js [hunk 12]
```diff
@@ -375,10 +432,7 @@ const VProMMsWayBbox = function (state = VProMMsWayBboxDefaultState, action) {
       state = _.cloneDeep(state);
       state.fetching = false;
       state.fetched = true;
-      state.bbox = action.json[Object.keys(action.json)[0]];
```
Comment (danielfdsilva) [inline app/assets/scripts/reducers/reducer.js]: Why did you go back with this?

**Pull request merged**